macaroonbakery/checker.py

```diff
@@ -36,4 +49,388 @@ def __hash__(self):
+        '''
+        nops = 0
+        for op in ops:
+            if op != LOGIN_OP:
+                nops += 1
+        if nops == 0:
+            raise ValueError('no non-login operations required in capability')
+
+        _, used = self._allow_any(ctx, ops)
+        squasher = _CaveatSquasher()
+        for i, is_used in enumerate(used):
+            if not is_used:
+                continue
+            for cond in self._conditions[i]:
+                squasher.add(cond)
+        return squasher.final()
+
+    def _check_conditions(self, ctx, op, conds):
+        declared = infer_declared_from_conditions(
+            conds,
+            self.parent.namespace())
+        ctx = context_with_operations(ctx, [op.action])
+        ctx = context_with_declared(ctx, declared)
+        for cond in conds:
+            err = self.parent._first_party_caveat_checker.\
+                check_first_party_caveat(ctx, cond)
+            if err is not None:
+                return None, err
+        return declared, None
+
+
+class AuthInfo(namedtuple('AuthInfo', 'identity macaroons')):
+    '''AuthInfo information about an authorization decision.
+    '''
+    __slots__ = ()
+
+    def __new__(cls, identity, macaroons):
+        '''
+        :param identity: holds information on the authenticated user as
+        returned identity_client. It may be None after a successful
+        authorization if LOGIN_OP access was not required.
+        :param macaroons: holds all the macaroons that were used for the
+        authorization. Macaroons that were invalid or unnecessary are
+        not included.
+        '''
+        return super(AuthInfo, cls).__new__(cls, identity, macaroons)
+
+
+class _CaveatSquasher(object):
+    def __init__(self, expiry=None, conds=None):
+        self._expiry = expiry
+        if conds is None:
+            conds = []
+        self._conds = conds
+
+    def add(self, cond):
+        if self._add(cond):
+            self._conds.append(cond)
+
+    def _add(self, cond):
+        try:
+            cond, args = parse_caveat(cond)
+        except ValueError:
+            # Be safe - if we can't parse the caveat, just leave it there.
+            return True
+
+        if cond == COND_TIME_BEFORE:
+            try:
+                et = pyrfc3339.parse(args)
+            except ValueError:
+                # Again, if it doesn't seem valid, leave it alone.
+                return True
+            if self._expiry is None or et <= self._expiry:
+                self._expiry = et
+            return False
+        elif cond in [COND_ALLOW, COND_DENY, COND_DECLARED]:
+            return False
+        return True
+
+    def final(self):
+        if self._expiry is not None:
+            self._conds.append(time_before_caveat(self._expiry).condition)
+        if len(self._conds) == 0:
+            return []
+        # Make deterministic and eliminate duplicates.
+        self._conds = sorted(set(self._conds))
+        prev = self._conds[0]
```

Comment:
I think that all this code is redundant because you've used set() above.

Reply:
done.